### PR TITLE
Document server middleware and interceptors

### PR DESCRIPTION
**Pull request by tobias-tengler**
### Changes

- Document middleware
- Document interceptors
- Document global state
- Document uploading files
- Better scrollbar for "In this article"

[Preview](https://chilly-server-docs.netlify.app/docs/hotchocolate/server)

### CODE REVIEW

File: website/src/docs/hotchocolate/server/index.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
+---
+title: Overview
+---
+
+In this section we will learn how we can configure our GraphQL server and look at some transport protocol related details.
```
Comment (michaelstaib) [inline website/src/docs/hotchocolate/server/index.md:5]:
```suggestion
In this section, we will learn how we can configure our GraphQL server and look at some transport protocol-related details.
```

File: website/src/docs/hotchocolate/server/index.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+# Middlewares
+
+Hot Chocolate comes with ASP.NET Core middlewares for accepting HTTP / WebSocket GraphQL requests, downloading the GraphQL schema or serving the [Banana Cake Pop](/docs/bananacakepop) GraphQL IDE.
```
Comment (michaelstaib) [inline website/src/docs/hotchocolate/server/index.md:9]:
```suggestion
Hot Chocolate comes with ASP.NET Core middlewares for accepting HTTP / WebSocket GraphQL requests, downloading the GraphQL schema, or serving the [Banana Cake Pop](/docs/bananacakepop) GraphQL IDE.
```

File: website/src/docs/hotchocolate/server/index.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+[Learn more about middlewares](/docs/hotchocolate/server/middlewares)
+
+# Interceptors
+
+Interceptors allow us to intercept GraphQL requests before they are being executed. There are interceptors for both GraphQL requests sent via HTTP as well as via WebSockets.
+
+In the case of WebSockets the interceptor also allows us to handle life cycle events such as when a client first connects.
```
Comment (michaelstaib) [inline website/src/docs/hotchocolate/server/index.md:17]:
```suggestion
In the case of WebSockets, the interceptor also allows us to handle life cycle events, such as when a client first connects.
```

File: website/src/docs/hotchocolate/server/interceptors.md [hunk 4]
```diff
@@ -0,0 +1,236 @@
+---
+title: Interceptors
+---
+
+Interceptors allow us to hook into protocol specific events. We can for example intercept an incoming HTTP request or a client connecting or disconnecting a WebSocket session.
```
Comment (michaelstaib) [inline website/src/docs/hotchocolate/server/interceptors.md:5]:
```suggestion
Interceptors allow us to hook into protocol-specific events. We can, for example, intercept an incoming HTTP request or a client connecting or disconnecting a WebSocket session.
```

File: website/src/docs/hotchocolate/server/interceptors.md [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+services
+    .AddGraphQLServer()
+    .AddHttpRequestInterceptor<HttpRequestInterceptor>();
+```
+
+If needed, we can also inject services into our custom `HttpRequestInterceptor` using its constructor.
+
+## OnCreateAsync
+
+This method is invoked for **every** GraphQL request sent via HTTP. It is a great place to set global state variables, extend the identity of the authenticated user or anything that we want to do on a per-request basis.
```
Comment (michaelstaib) [inline website/src/docs/hotchocolate/server/interceptors.md:38]:
```suggestion
This method is invoked for **every** GraphQL request sent via HTTP. It is a great place to set global state variables, extend the identity of the authenticated user or do anything that we want to do on a per-request basis.
```

File: website/src/docs/hotchocolate/server/index.md (hunk 2 above)
Comment (benmccallum) [reply website/src/docs/hotchocolate/server/index.md:9]: I believe the plural of middleware is just middleware. Just like software --> software. 

Worth double checking but it feels weird to say middlewares. It's in the title and this sentence. Maybe other places.

File: website/src/docs/hotchocolate/server/middlewares.md [hunk 6]
```diff
@@ -0,0 +1,372 @@
+---
+title: Middlewares
```
Comment (benmccallum) [inline website/src/docs/hotchocolate/server/middlewares.md:2]:
```suggestion
title: Middleware
```

File: website/src/docs/hotchocolate/server/middlewares.md [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+---
+
+In Order for Hot Chocolate to be accessible using a HTTP endpoint, we need to register some middlewares with ASP.NET Core.
```
Comment (benmccallum) [inline website/src/docs/hotchocolate/server/middlewares.md:5]:
```suggestion
In Order for Hot Chocolate to be accessible using a HTTP endpoint, we need to register some middleware with ASP.NET Core.
```

**Review by benmccallum — COMMENTED** [review]
Just adding some comments on a drive but didn't get through it all sorry. 

Was trying to make it clearer that an interceptor is basically hooked up and then can run code at certain events rather than how it's worded in some places now that make it seem like it just intercepts before execution. Hopefully that makes sense. 

Gotta run but don't wait for me to merge this by any means, just some quick thoughts :) nice work as usual!

File: website/src/docs/hotchocolate/server/interceptors.md [hunk 8]
```diff
@@ -0,0 +1,236 @@
+---
+title: Interceptors
+---
+
+Interceptors allow us to hook into protocol specific events. We can for example intercept an incoming HTTP request or a client connecting or disconnecting a WebSocket session.
+
+# IHttpRequestInterceptor
+
+Each GraphQL request sent via HTTP can be intercepted using an `IHttpRequestInterceptor` before it is being executed. Per default Hot Chocolate registers a `DefaultHttpRequestInterceptor` for this purpose.
```
[the remark this answers is not in the document]
Comment (tobias-tengler) [reply website/src/docs/hotchocolate/server/interceptors.md:9]: Yes, there's indeed only the OnCreateAsync.

**Comment by arif-hanif** [comment]
@tobias-tengler  this is great!

**Review by arif-hanif — APPROVED** [review]
approved

**Pull request merged**